File: Scr_Java/modelo/Torre.java
package modelo;

// Base comun de todas las torres. Cada subclase define sus stats iniciales
// (en el constructor) y su propia progresion de mejora en mejorar().
public abstract class Torre {
    protected int dano;
    protected int alcance;
    protected final int costo;
    protected int nivel;
    protected final Coordenada posicion;

    protected Torre(int dano, int alcance, int costo, Coordenada posicion) {
        this.dano = dano;
        this.alcance = alcance;
        this.costo = costo;
        this.posicion = posicion;
        this.nivel = 1;
    }

    // Una torre solo puede atacar objetivos dentro de su alcance (distancia Manhattan).
    public boolean estaEnRango(Coordenada posEnemigo) {
        return posicion.distanciaA(posEnemigo) <= alcance;
    }

    public void atacar(Cozy objetivo) {
        objetivo.recibirDano(dano);
    }

    // Cada subclase define su propio incremento de dano/alcance al mejorar.
    public abstract void mejorar();

    public abstract String getTipo();

    // Revierte una mejora anterior. Usado exclusivamente por
    // ComandoMejorarTorre.deshacer() para volver al snapshot previo.
    public void restaurar(int nivel, int dano, int alcance) {
        this.nivel = nivel;
        this.dano = dano;
        this.alcance = alcance;
    }

    public int getDano() {
        return dano;
    }

    public int getAlcance() {
        return alcance;
    }

    public int getCosto() {
        return costo;
    }

    public int getNivel() {
        return nivel;
    }

    public Coordenada getPosicion() {
        return posicion;
    }
}

File: cpp/modelo/Torre.cpp
#include "Torre.h"

// Guarda los stats iniciales que le pasa la subclase y arranca en nivel 1.
Torre::Torre(int danio_, int alcance_, int costo_, const Coordenada& posicion_)
    : danio(danio_), alcance(alcance_), costo(costo_), nivel(1), posicion(posicion_) {}

// La torre alcanza al enemigo si la distancia entre ambas casillas es
// menor o igual que su alcance actual.
bool Torre::estaEnRango(const Coordenada& posEnemigo) const {
    return posicion.distanciaA(posEnemigo) <= alcance;
}

// Simplemente le pasa el danio de la torre al metodo recibirDano() del enemigo.
void Torre::atacar(Cozy& objetivo) const {
    objetivo.recibirDano(danio);
}

// Sobrescribe nivel/danio/alcance con los valores que le pasan (el
// "snapshot" que guardo ComandoMejorarTorre antes de mejorar).
void Torre::restaurar(int nivel_, int danio_, int alcance_) {
    this->nivel = nivel_;
    this->danio = danio_;
    this->alcance = alcance_;
}

int Torre::getDano() const { return danio; }
int Torre::getAlcance() const { return alcance; }
int Torre::getCosto() const { return costo; }
int Torre::getNivel() const { return nivel; }
const Coordenada& Torre::getPosicion() const { return posicion; }